meta-sokol-flex-common/classes/image-root-password.bbclass

```diff
@@ -20,13 +22,29 @@ allow root login with an empty password."
 EMPTY_ROOT_PASSWORD = "${@'empty-root-password' if d.getVar('ROOT_PASSWORD', True) == '0' else ''}"
 IMAGE_FEATURES += "${EMPTY_ROOT_PASSWORD}"
 
-inherit extrausers
+def encrypt_root_pw(d):
+    import os
+    import crypt
+    import base64
+
+    prefix = '$6$'
+
+    password = d.getVar('ROOT_PASSWORD')
+    if any(password == pattern for pattern in [ 'None', '0', '*' ]):
+        return ''
 
-# This variable indirection allows for the possibility of programmatically
-# generating the root password, if so desired, without mucking up bitbake's
-# variable checksums.
-ACTUAL_ROOT_PASSWORD = "${ROOT_PASSWORD}"
-EXTRA_USERS_PARAMS:prepend = "${@'usermod -P \'${ACTUAL_ROOT_PASSWORD}\' root;' if d.getVar('ACTUAL_ROOT_PASSWORD') not in ['', '0', '*'] else ''}"
+    salt = base64.b64encode(os.urandom(8))
+
+    rounds = d.getVar('ROOT_PASSWORD_HASH_ROUNDS')
+    if rounds is not None:
+        rounds = max(1000, min(999999999, int(rounds) or 10000))
+        prefix += 'rounds={0}$'.format(rounds)
+
+    return crypt.crypt(password, prefix + salt.decode('UTF-8')).replace('$','\$')
+
+inherit extrausers
+ACTUAL_ROOT_PASSWORD = '${@encrypt_root_pw(d)}'
+EXTRA_USERS_PARAMS:prepend = "${@'usermod -p \'${ACTUAL_ROOT_PASSWORD}\' root;' if d.getVar('ACTUAL_ROOT_PASSWORD') != '' else ''}"
```

Comment:
As far as I know, his is going to expand ACTUAL_ROOT_PASSWORD regardless of the logic, due to how variable expansion happens, though it won't be *used* in all cases. Still, no point running `encrypt_root_pw` if we don't have to. You should fix this to be `'usermod -p \'%s\' root;' % d.getVar('ACTUAL_ROOT_PASSWORD')`.

Reply:
What I'm suggesting is this:
```
def encrypt_root_pw(d):
    import os
    import crypt
    import base64
    import shlex

    prefix = '$6$'

    password = d.getVar('ROOT_PASSWORD')
    if any(password == pattern for pattern in [ 'None', '0', '*' ]):
        return ''

    salt = base64.b64encode(os.urandom(8))

    rounds = d.getVar('ROOT_PASSWORD_HASH_ROUNDS')
    if rounds is not None:
        rounds = max(1000, min(999999999, int(rounds) or 10000))
        prefix += 'rounds={0}$'.format(rounds)

    return shlex.quote(crypt.crypt(password, prefix + salt.decode('UTF-8')).replace('$','\$'))

inherit extrausers
ACTUAL_ROOT_PASSWORD = '${@encrypt_root_pw(d)}'
EXTRA_USERS_PARAMS:prepend = "${@'usermod -p %s root;' % d.getVar('ACTUAL_ROOT_PASSWORD') if d.getVar('ACTUAL_ROOT_PASSWORD') != '' else ''}"
```

Reply:
Yeah, I was confused for the usermod line; from the previous comment it looked like that an empty string case won't be handled.